[SIP-29] Add support for row-level security

altef opened the pull request:
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Many BI applications, particularly in multi-tenancy scenarios, require support for row-level security. That is, the ability to show different slices of a table to users based on some user attribute.   To accomplish this, I've added a new model to describe row level security filters, which references a Table and a Role. So when adding a row level security filter, you specify a particular Role and Table.

When querying that table, the applicable filters are added to the query.  I've modified the query function [here](https://github.com/altef/incubator-superset/commit/ab6ae456daecd34e676273fa4d43c52875e00e94#diff-6506def3966137541a05177a6fb169d7R981-R987) to add any relevant to the WHERE clause.

As well, I've added a UI for managing the row level security filters.  And for convenience, added it as a related view for tables.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:

 ![menu-before](https://user-images.githubusercontent.com/4316295/69895421-dba32600-12e3-11ea-85fb-bfbffd36abd0.JPG)

After:

1. Row level security filters added to the Security menu:
![menu](https://user-images.githubusercontent.com/4316295/69893637-5743a980-12c9-11ea-8020-1f87f0c46c5d.jpg)

2. Row level security list
![list](https://user-images.githubusercontent.com/4316295/69893640-6c203d00-12c9-11ea-9962-81a320ffc2cd.jpg)

3. Row level security interface
![edit](https://user-images.githubusercontent.com/4316295/69893642-76dad200-12c9-11ea-8070-22f6f5f1d8cc.jpg)

That allows for the management of the RowLevelSecurityFilters model.  Additionally, for convenience, I've added it as a related view for tables.

![table-related-view](https://user-images.githubusercontent.com/4316295/69893652-a25dbc80-12c9-11ea-9f57-d2ec6b609315.jpg)

After logging in with a user assigned to that role, I can still supply additional filters:

![additional-filters](https://user-images.githubusercontent.com/4316295/69893666-c91bf300-12c9-11ea-9504-6c1a20727307.jpg)

The generated SQL includes the additional filters AND the clause supplied by the row level security filter(s).

![generated-sql](https://user-images.githubusercontent.com/4316295/69893678-e6e95800-12c9-11ea-82f6-5fe9160a7510.jpg)

### TEST PLAN
<!--- What steps should be taken to verify the changes -->
Everything seems to be working as expected on my end, but a few things should be done to verify the changes.

1. Set up a limited user
   1. Create a new role and give it access to a table.
   2. Create a new user and assign it the ``Gamma`` role, as well as the role you've just created.
   3. Create a ``Row level security filter`` and assign it the table and row.
2. Ensure that the table is still working as expected for you (there should be no change)
3. Create a simple dashboard built on that table
4. Verify the limited user is only seeing the filtered data 
   1. Log in as the limited user
   2. Check the table to ensure the filter is being applied.
   3. Check the dashboard to ensure the data is being limited.

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [x] Has associated issue: [#8644 ](https://github.com/apache/incubator-superset/issues/8644)
- [x] Changes UI
- [x] Requires DB Migration. 
    - To add a new model.)
- [ ] Confirm DB Migration upgrade and downgrade tested. 
    - I tested it, but presumably this doesn't  apply to me?
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

altef committed:
Support and apply filters.

altef committed:
Added the UI for row level security, and moved it all under SQLA in order to access the Table model more easily.

altef committed:
Added a row level security filter documentation entry.

altef committed:
Accidentally added two new lines to this file.

altef committed:
Blacked and iSorted, hopefully.  Also, sometimes g.user may not be set.

altef committed:
Another isort, and handling g not having a user attribute another way.

altef committed:
Let's try this again #CI tests.

altef committed:
Adjusted import order for isort; I was sure I'd already done this..

villebro commented:
Great initiative! I looked at the PR for this SIP and a few things came to mind:

To address the issue you raise in the docs where belonging to two departments results in zero rows (`WHERE dept = 1 AND dept = 2`), I propose changing the `Clause` field to JSON, putting each clause behind a key, which are later ORed together. Example: There are four RLS filters specified for a table: two for departments ("Finance", "Risk") and two for time history ("30 days", "90 days").
"Finance" role:
```json
{
    "dept": "dept_id = 'Finance'"
}
```
"Risk" role:
```json
{
    "dept": "dept_id = 'Risk'"
}
```

"30 days" role:
```json
{
    "history": "report_date >= current_timestamp() - 30"
}
```

"90 days" role:
```json
{
    "history": "report_date >= current_timestamp() - 90"
}
```

Now, if a user belonged to all four groups, this would result in the following WHERE clause:
```sql
((dept_id = 'Finance') OR (dept_id = 'Risk')) AND 
((report_date >= current_timestamp() - 30) OR (report_date >= current_timestamp() - 90))
```

In this case the user would see 90 days of history, and also all rows for Finance AND Risk.

Also, I think it would be good to be able to define a default WHERE clause if a user doesn't belong to any of the specified groups.

altef committed:
Row level filters should be wrapped in parentheses in case one contains an OR.

altef committed:
Oops, did not think that would change Black's formatting.

altef commented:
Hey - thanks for looking it over!  I wanted to point that out in the docs for clarity, 
and viewed it as more of something the user should be aware of than as an issue.  I can definitely 
see a certain use-case where it may not be ideal, but am of the mindset that it's better
to show no data (which can be easily detected and fixed), should the user introduce a 
conflict in the filters, than to be overly permissive and allow a user to see data they otherwise
shouldn't be able to see.  Should _that_ occur, it may not be readily apparent.

I think everything you've mentioned can be mostly accomplished as-is, by taking a slightly 
different approach to the role/filter composition.

1.  Filters clauses can contain OR.  You could simply make a filter with the 
clause `dept_id = 'Finance' OR dept_id = 'Risk'` and assign it to a particular role.  Obviously 
this is not ideal if you want to assign a user to more than one department by assigning them 
two roles.  What I'm not sure of is how much this needs to come up.  If someone has access to 
both the `Finance` and the `Risk`, are they really in each department, or are they some 
higher-up who should have greater access? The current method, I hope, seems analogous to superset's **Restricting access to a subset of data sources** recommendation.

Which leads me to realize that there was a bug in my initial implementation - filters are now 
wrapped in parentheses in case one has an OR in it.

2.  To reference the **Restricting access to a subset of data sources** again, it's recommended 
to assign a limited user to the gamma role, and to make a role giving them access to the 
datasource in question.  In the case of creating a default filter, you could just assign a 
filter to that role.  However, it is not exactly what you have in mind, since it wouldn't 
be _overwritten_ by additional filters.


My worry is that we'd be increasing complexity, in the code, as well as in ease-of-understanding on the 
user's part, for minimal gains.  However, it could be that I'm taking the wrong approach here.  
Is this something others would find useful?



I would like to eventually add (probably as a separate SIP?) arbitrary user attributes that 
can then be referenced in the row level security filters.  But I'd want to talk to someone 
about that, as I have many questions in that regard.

villebro commented:
I think it is important to support belonging to multiple roles early on. Think AD/LDAP in a corporate setting; not uncommon to belong to hundreds of groups. Regarding implementation, I would propose just adding a column "role_based_filters"  or similar to the tables table with the metadata:
```json
{ 
   "defaults":{ 
      "duration":"report_date >= current_timestamp() - 1",
      "dept":"false"
   },
   "roles":{ 
      "finance":{ 
         "duration":"report_date >= current_timestamp() - 30",
         "dept":"dept_id = 1"
      },
      "risk":{ 
         "dept":"dept_id = 2"
      }
   }
}
```
In this example, users that don't belong to any group would get a WHERE clause that returns zero rows due to the "false" clause (`SELECT col FROM table WHERE FALSE` -> no rows), and by default only the last days data would be available. If the user belongs to the "risk" Role, they would see only "dept_id = 2" for the last day (default clause for "duration"), whereas "finance" would see "dept_id = 1" for the last 30 days. Belonging to both would return data for both departments with 30 days of data.

One could later add the same column to the charts table, making it possible to introduce the same functionality on a per chart basis. With regards to the filter statements, I would propose using the same filter format that's currently used for `adhoc_filters`, which would enable us to leverage existing React components that allow for a much more user friendly means to add filters. To introduce the functionality, I would break the SIP into two parts; first introducing the backend functionality, i.e. adding the new column to table, making it possible to edit the filters by poking at the table metadata, and later adding proper UI functionality for editing the metadata.

justin-barton commented:
> I think it is important to support belonging to multiple roles early on. Think AD/LDAP in a corporate setting; not uncommon to belong to hundreds of groups. Regarding implementation, I would propose just adding a column "role_based_filters" or similar to the tables table with the metadata:
> 
> ```json
> {
>   "dept": {
>     "default": "false",
>     "roles": {
>       "finance": "dept_id = 1",
>       "risk": "dept_id = 2"
>     }
>   },
>   "duration": {
>     "default": "report_date >= current_timestamp() - 1",
>     "roles": {
>       "finance": "report_date >= current_timestamp() - 30"
>     }
>   }
> }
> ```
> 
> In this example, users that don't belong to any group would get a WHERE clause that returns zero rows due to the "false" clause (`SELECT col FROM table WHERE FALSE` -> no rows), and by default only the last days data would be available. If the user belongs to the "risk" Role, they would see only "dept_id = 2" for the last day (default clause for "duration"), whereas "finance" would see "dept_id = 1" for the last 30 days. Belonging to both would return data for both departments with 30 days of data.
> 
> One could later add the same column to the charts table, making it possible to introduce the same functionality on a per chart basis. With regards to the filter statements, I would propose using the same filter format that's currently used for `adhoc_filters`, which would enable us to leverage existing React components that allow for a much more user friendly means to add filters. To introduce the functionality, I would break the SIP into two parts; first introducing the backend functionality, i.e. adding the new column to table, making it possible to edit the filters by poking at the table metadata, and later adding proper UI functionality for editing the metadata.

@villebro I've been looking through your suggestions and I believe that there are a number of potential bugs / unintended consequences:

1. When someone belongs to multiple groups, setting the default behaviour to OR together all the permissions seems like a dangerous assumption. This errs on the side of being maximally permissive, which is usually not advisable. The current setup forces the user to define what happens at intersections by creating another role for any combined roles and explicitly setting the clauses.

2. Even if we were to OR together permissions, using a key-based JSON approach would likely not be the right one. By way of example, let's say that on table `exports`: 
```
Role A has filters: country='Freedonia' AND item='Apples'
Role B has filters: country='Ruritania' AND item='Oranges'
```
For someone that has Role A and Role B, ORing together at a key-level would give the user access to not only reports on Freedonia Apple exports and Ruritania Orange exports (as you might expect) but also Ruritania Apple exports and Freedonia Orange exports (which may not be expected/desired in a lot of use cases).

3. Having users directly manage large JSON objects to control row-based security is unduly cumbersome and inaccessible to some users, so a UI / SQL approach is preferable.

4. The concept of defaults (and how to use them) again requires assumptions that will not fit a lot of use cases. For example, your proposed setup only works where the default is more restrictive than all of the roles. If there were a case where the default was to show last 15 days (`report_date >= current_timestamp() - 15`) but one of the roles was to restrict to only seeing the most recent day's data (`report_date >= current_timestamp() - 1`) then your setup would actually escalate permissions.

5. The proposal to break the SIP into two parts, the first being only a back-end change to tables, means that most users will not have use of row-level security when this is released, which defeats the purpose of the SIP.


In answer to the above, I would suggest one of the following two approaches:

A. Accept the pull request as-is, continuing to force users to explicitly define the permissions for any intersection cases (since there aren't any defaults that will work in all scenarios and that would lead to predictable behaviour for the user)

B. Modify the current pull request to allow for roles to be combined by ORing together the entire clause in brackets, e.g.: 
`Role A & Role B: WHERE (country='Freedonia' AND item='Apples') OR (country='Ruritania' AND item='Oranges')`

With a preference for A.

villebro commented:
Thanks for the feedback, very good to have an exhaustive discussion prior to committing to any approach.

I think there might be a misunderstanding about how the ORing approach should work. The basic idea is this:

1) AND across different keys
2) OR within keys
3) apply default if user doesn't belong to any roles within a certain key

In my example, the following WHERE clauses would be generated:

1) doesn't belong to any groups: `((false)) AND ((report_date >= current_timestamp() - 1))`
2) belongs to "finance": `((dept_id = 1)) AND ((report_date >= current_timestamp() - 30))`
3) belongs to "risk": `((dept_id = 2)) AND ((report_date >= current_timestamp() - 1))`
3) belongs to "finance" and "risk": `((dept_id = 1) OR (dept_id = 2)) AND ((report_date >= current_timestamp() - 30))`

For your example case, providing that these filter groups were made with the same key (as I understand they should), the following WHERE clause would be generated: `((country='Freedonia' AND item='Apples') OR (country='Ruritania' AND item='Oranges'))`, i.e. the user would *not* see Ruritania Apple exports.

In the case of having a default duration of 30 days and specifying a more restrictive filter group for 1 day, the default would not be applied, i.e. the user would only see one day's worth of data. Of course, if the user belonged to *two* restrictive groups, 15 days and 1 days, the more permissive role of 15 days would in practice apply. However, this seems like a logical error in how the roles are assigned to users.

The proposal to start by rolling out the backend functionality was merely a proposal to keep the PRs as small as possible and easier to review/develop. However, I'm sure they can be done together, assuming the person working on the PR is proficient in both the frontend and backend aspects of the codebase.

justin-barton commented:
> Thanks for the feedback, very good to have an exhaustive discussion prior to committing to any approach.
> 
> I think there might be a misunderstanding about how the ORing approach should work. The basic idea is this:
> 
> 1. AND across different keys
> 2. OR within keys
> 3. apply default if user doesn't belong to any roles within a certain key
> 
> In my example, the following WHERE clauses would be generated:
> 
> 1. doesn't belong to any groups: `((false)) AND ((report_date >= current_timestamp() - 1))`
> 2. belongs to "finance": `((dept_id = 1)) AND ((report_date >= current_timestamp() - 30))`
> 3. belongs to "risk": `((dept_id = 2)) AND ((report_date >= current_timestamp() - 1))`
> 4. belongs to "finance" and "risk": `((dept_id = 1) OR (dept_id = 2)) AND ((report_date >= current_timestamp() - 30))`
> 
> For your example case, providing that these filter groups were made with the same key (as I understand they should), the following WHERE clause would be generated: `((country='Freedonia' AND item='Apples') OR (country='Ruritania' AND item='Oranges'))`, i.e. the user would _not_ see Ruritania Apple exports.
> 
> In the case of having a default duration of 30 days and specifying a more restrictive filter group for 1 day, the default would not be applied, i.e. the user would only see one day's worth of data. Of course, if the user belonged to _two_ restrictive groups, 15 days and 1 days, the more permissive role of 15 days would in practice apply. However, this seems like a logical error in how the roles are assigned to users.
> 
> The proposal to start by rolling out the backend functionality was merely a proposal to keep the PRs as small as possible and easier to review/develop. However, I'm sure they can be done together, assuming the person working on the PR is proficient in both the frontend and backend aspects of the codebase.

I think that I'm following. So in your framework, my example would look something like:
```
{
  "somekey": {
    "default": "false",
    "roles": {
      "freedonia-orchards": "country='Freedonia' AND item='Apples'",
      "ruritania-groves": "country='Ruritania' AND item='Oranges'"
    }
  }
}
```
I believe that functionally this is very similar to the current PR, with the following difference:
- In the current PR, combinations of roles must be created as new roles and the filter clauses explicitly defined
- In your proposed setup the logic for how the clauses within roles should be combined is explicitly defined by the user at the time of creation in the AND/OR hierarchy of the JSON structure

Is that accurate? How would you envision making this available to end-users in the UI?

mistercrunch commented on superset/connectors/sqla/models.py:
```diff
@@ -977,6 +977,17 @@ def _get_top_groups(
+        filters = security_manager.get_row_level_security_filters(
+            self
+        )  # Self is the Table model
+        if len(filters) > 0:
```
`if filters:` is the pythonic way

altef replied:
removed

mistercrunch commented on superset/connectors/sqla/models.py:
```diff
@@ -977,6 +977,17 @@ def _get_top_groups(
         return or_(*groups)
 
     def query(self, query_obj: Dict) -> QueryResult:
+        filters = security_manager.get_row_level_security_filters(
```
This new block should probably be in it's own method `_get_row_level_filters`

mistercrunch replied:
Also wondering if this logic should be in `get_sqla_query` instead (unclear). One thing that would be neat would be to enable templating in the text, and if this was done in `get_sqla_query` you'd have a handle on `template_processor.process_template` in that context. 

Eventually it'd be great to be able to base row level rules on user attribute as in `WHERE sales_team_id = {{ current_user.extras.sales_team_id }}`, (note that `extras` doesn't really exist yet) instead of having to create entries for each sales team. You could even imagine things like `WHERE country IN {{ call_external_service.list_access_to_countries(current_user.id) }}`

altef replied:
Moved to within `get_sqla_query`, using `text`.  I'd like the user attributes very much, eventually.

mistercrunch commented on superset/connectors/sqla/models.py:
```diff
@@ -977,6 +977,17 @@ def _get_top_groups(
+        if len(filters) > 0:
+            where = query_obj["extras"]["where"]
+            if len(where) > 0:
+                where = "({}) AND ".format(where)
```
Would be better to use the SQLA expression toolkit. `from sqlalchemy import and_`, `from sqlalchemy.sql import text`

altef replied:
Swapped

mistercrunch commented on superset/connectors/sqla/models.py:
```diff
@@ -1169,3 +1180,20 @@ def get_extra_cache_keys(self, query_obj: Dict) -> List[Any]:
+    __tablename__ = "row_level_security_filters"
+    id = Column(Integer, primary_key=True)  # pylint: disable=invalid-name
+    role_id = Column(Integer, ForeignKey("ab_role.id"), nullable=False)
+    role = relationship(
```
Shouldn't this be a many to many?

altef replied:
It's now many to many

mistercrunch reviewed (changes requested):
This is not too far from where this should be, but I'm thinking there are some important things to consider here. I think the many-to-many to role is important. 

Another point is who can edit these rules and making sure security works well out of the box. Only admins should be able to edit these...

villebro commented:
@justin-barton yes, precisely like that. I refined my original proposal somewhat, and concluded that the default value should be present in the OR clause. For instance, if you would want to give access to everyone for pear sales in Megalomania, the key-value pair `"default": "country='Megalomania' AND item='Pears'"` should be present under `"somekey"` to ensure users in the `ruritania-groves` Role aren't left out.

My main motivation here is to ensure we can construct arbitrarily complex role based filters, and after this last amendment I believe this is pretty solid. So the basic premise is:
- Default value for key -> what everyone should see
- Additional keys -> increased security
- Additional roles within a key -> relaxed security,

which I feel should be a pretty good trade-off between complexity and versatility.

@mistercrunch @dpgaspar do you feel this should be done using the FAB model, or as a column under table? I was thinking one could create a React component for editing the groups/roles, that in turn leverages the existing `AdhocFilterEditPopover` React component.

altef committed:
Changes as per @mistercrunch.

altef committed:
RLS filters are now many-to-many with Roles.

altef committed:
Updated documentation to reflect RLS filters supporting multiple rows.

altef committed:
Let's see what happens when I set it to the previous revision ID

altef committed:
Merge branch 'master' into row-level-security

altef committed:
Updated from upstream.

altef committed:
There was a pylint error.

altef commented:
@mistercrunch thanks!  RLS filters should now be many-to-many to roles.  I've moved the logic to `get_sqla_query`, using SQLA expression toolkit's `text`.

Here's a screenshot of the updated RLS filters UI supporting multiple roles.
![image](https://user-images.githubusercontent.com/4316295/70167974-be23e280-167c-11ea-9e4f-f6f3cd69c489.png)

I agree completely about including user attributes - that'll be such a neat and useful feature.

mistercrunch commented:
Yesterday I went straight to the code and missed much of the conversation and comments here. I think quickly things become better handled based on data over metadata. By that I mean if there are complex rules over which sales rep can see which regions, or even just if there's a high-ish cardinality of rules (say over 10), things are best represented as a subquery based on local data.

That's easily expressed as something like `region IN (SELECT region FROM rules_table WHERE user_id = {{current_user.id}})`. Then a single rule will do and it becomes about managing the `rules_table`.

I'm not against a more structured format as in (field, operator, value), as long as value is templatable and operator has a subquery option. It does seem logic to go logical `AND` across fields and logical `OR` (or `IN`) on same field.

villebro commented:
It seems there is general consensus to go with the originally proposed approach i.e. creating a new table with one row per rule over adding a column in the tables table with JSON (the one I suggested). Unless merged by then, I can do a review pass on the code in the coming days so we can get this functionality in as soon as possible.

mistercrunch commented on superset/connectors/sqla/models.py:
```diff
@@ -834,6 +843,10 @@ def get_sqla_query(  # sqla
                         where_clause_and.append(col_obj.get_sqla_col() != None)
+
+        where_clause_and += [
+            text("({})".format(f)) for f in self._get_row_level_filters()
```
I think sqlalchemy should be doing the right thing here around parenthesis, so no need to add them explicitely
https://docs.sqlalchemy.org/en/13/core/sqlelement.html#sqlalchemy.sql.expression.BinaryExpression.self_group

Something like 
```python
where_clause_and += [text(f) for f in self._get_row_level_filters()]
```
Oh and I guess the `text()` call could be in the method, could rename to `_get_sqla_row_level_filters`, so it's just `where_clause_and += self._get_row_level_filters()`

altef replied:
I tested this by adding an RLS filter that contained an `OR`.  It didn't end up getting wrapped in parentheses. So I still have the `"({})".filter(f)` in there, but I've moved it to the `_get_sqla_row_level_filters` method.
It looks like similar was done in the `get_sql_query()` method, presumably for similar reason
https://github.com/apache/incubator-superset/blob/cae0583e0fbf8b60b9ee86179b4defef5da9b2c2/superset/connectors/sqla/models.py#L841

mistercrunch commented:
It's important to note that we'll be killing the old "table editor" eventually in favor of the React-based one that's accessible in the explore view. That means we'll have to implement the RLS UI there prior to deprecation. That brings the question: "where is the best place to manage RLS rules? should it be in the table editor? in the role editor? ...".

One more comment about adding more semantics around clauses (the idea of the JSON blob there). To me it's not necessary for MVP, and I think the the forced logical `AND` is easy to express and reason about. I get how it'd be nice to have the logical OR on columns eventually, but asking for users to write JSON in a UI is not great, clearly a pattern we're moving away from. Maybe as we move to the new table editor we can design the UI that allows for that logical OR idea.

Few more things, by default users shouldn't be able to modify RLS settings. For now the way to do that is to add the model view here: https://github.com/apache/incubator-superset/blob/master/superset/security.py#L100-L109 , I think that should do it.

The last thing is using the template processor. I think that's easy to do, just a matter of passing the template processor to `_get_row_level_filters`.

villebro reviewed (changes requested):
I did a quick pass over the code and tested locally. A few comments:

1) a new migration has been merged to master since this branch was cut. You will need to rebase and make the `down_revision` point to `89115a40e8ea`.
2) Templating is currently not being applied. If you check a few rows below the part where the `where_clause_and +=` list comprehension is defined in this PR you can see how to add templating (the possibitility of doing an `IN` subquery based on `{{current_user.id}}` @mistercrunch was referring to). 
3) Currently RLS filters are not considered when calculating the cache key, which will cause results to leak from one person to the other if caching is enabled. Check the `cache_key` method in `viz.py` and `common/query_object.py`; I propose adding a key `rls_keys` or similar that contains all the RLS filter ids that are present in the WHERE clause.
4) a unit test would be good, especially to guard against views leaking from one role to another.

I think it is vital to fix the caching bug prior to merging, and adding templating support would be an added bonus. Based on my testing, this fits well in a use case where the intent is to start with unrestricted access, but additively restrict access to certain rows for certain roles. However, this doesn't address a more common need I usually see, which starts with no access, unless explicitly given via a role. However, I think this is a good starting point, and I'm sure we can add functionality later on when this gets refactored to the React-based table editor.

altef committed:
Added RLS ids to the cache keys; modified documentation; added template processing to RLS filters.

altef committed:
Merge branch 'master' into row-level-security

altef committed:
A new migration was merged in.

altef commented on docs/security.rst:
```diff
@@ -153,3 +153,28 @@ a set of data sources that power dashboards only made available to executives.
+``date_field > DATE_SUB(NOW(), INTERVAL 30 DAY)``.  It can also support multiple 
+conditions: ``client_id = 6 AND advertiser="foo"``, etc. 
+
+You can throw whatever you want in there to define the subset of the table you want the roles in question to have access to.
```
removed

altef committed:
Removed RLS cache key from query_object.

altef commented:
* Added template processing and tested it with the RLS clause `advertiser_name = "{{ current_username() }}"`
* Added RLS IDs to the `viz.py` cache key, but removed it from `common/query_object.py`.  Could you tell me a little more about this one? How does it get called? (I'm trying to figure out how to trigger it for testing.)  Additionally, it doesn't seem to reference a datasource, though possibly I am blind.
* Updated the docs to eliminate the contentious line

villebro commented:
@altef check `superset/common/query_context.py`: https://github.com/apache/incubator-superset/blob/master/superset/common/query_context.py#L163-L171 . Here you have access to the datasource, and you need to make sure that the RLS keys are passed to the `cache_key` method in `query_object`.

altef committed:
RLS added to the cache_key from query_context.

altef commented:
@villebro Awesome, thanks!

villebro reviewed (approved):
LGTM; however, you need to rebase as there's a conflict on `security.py`

toop commented:
test error:
        AttributeError: 'SupersetSecurityManager' object has no attribute 'get_row_level_security_filters'

altef committed:
Merge branch 'master' into row-level-security

toop commented:
I found error @altef ：
Use filter item='Oranges' it·s worked，But use  filter item="Oranges" it`s  Oracle error: ora-00904: "Oranges": invalid identifier

altef commented:
@toop I think (if memory serves) single quotes are used to enclose literals in Oracle.  Using double quotes in that situation is probably resulting in invalid SQL.  Allowing arbitrary SQL clauses lends a lot of power and flexibility but opens the door to user-error, which is hopefully reasonable since only admins can set the RLS filters.

villebro commented:
@toop I agree with @altef; in Oracle, surrounding a string with double quotes implies that it's referencing a database object, in this case the value of the case sensitive column "Oranges", which is consistent with the error message (ora-00904).

@mistercrunch are you ok with the latest changes? If so, I think we're ready to merge.

etr2460 commented on superset/migrations/versions/0a6f12f60c73_add_role_level_security.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
nit, remove the Alembic comments

etr2460 commented on superset/migrations/versions/0a6f12f60c73_add_role_level_security.py:
```diff
@@ -0,0 +1,65 @@
+        "rls_filter_roles",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("role_id", sa.Integer(), nullable=False),
+        sa.Column("rlsfilter_id", sa.Integer(), nullable=True),
```
probably should have an underscore between rls and filter

etr2460 commented on superset/security/manager.py:
```diff
@@ -877,3 +878,39 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+                for f in table.row_level_security_filters
+                if any(r.id in roles for r in f.roles)
+            ]
+            ids.sort()
```
why do we need to sort here? It's not clear to me

altef replied:
It's used to add to the cache key, where presumably we care which roles they're assigned rather than the order they've been assigned them.  The sort could be done where the RLS IDs are added to the cache keys, but then it'd be in two places rather than one.

etr2460 reviewed (changes requested):
can you add testing in SQL Lab to your test plan?

Also, please add a section in `UPDATING.md` about this major feature

altef committed:
Changes as per @etr2460.

altef committed:
Updating entry for RLS pull request.

altef committed:
Merge branch 'master' into row-level-security

altef committed:
Another migration to skip.

etr2460 reviewed (approved):
The changes with regards to my comments look good, so i'll unblock now. Sorry for the delay over the holidays!

serenajiang commented on superset/security/manager.py:
```diff
@@ -877,3 +878,39 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+            ]
+            ids.sort()  # Combinations rather than permutations
+            return ids
+        except AttributeError:
```
Hmm, I know it's a bit more verbose, but I prefer checking for the existence of the attribute using `hasattr` rather than catching the `AttributeError`.

altef replied:
Isn't `try/except` more "pythonic"?  I have no problem changing this, but worry about someone else then requesting it be changed back.  Can we get a second opinion on this one?

villebro replied:
I agree with @serenajiang here; best to be explicit about the attribute we`re after.

altef replied:
done

serenajiang commented on superset/security/manager.py:
```diff
@@ -877,3 +878,39 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+        :returns: A list of clause strings.
+        """
+        try:
+            roles = [role.id for role in g.user.roles]
```
Lots of the same code as `get_rls_filters` - maybe stick this in a generalized method?

altef replied:
but.. This is `get_rls_filters`, isn't it?

serenajiang replied:
My bad - meant `get_rls_ids` XD

altef replied:
generalized

serenajiang commented on superset/migrations/versions/0a6f12f60c73_add_role_level_security.py:
```diff
@@ -0,0 +1,61 @@
+def upgrade():
+    op.create_table(
+        "row_level_security_filters",
+        sa.Column("created_on", sa.DateTime(), nullable=True),
```
All these columns should be in the model.

altef replied:
I think they are in the model via the `mixin`s.  Am I wrong about this?  These migrations were automatically created using `superset db migrate` as per the contributing readme

serenajiang replied:
Ah, forgot about that. Nvm.

serenajiang commented on superset/security/manager.py:
```diff
@@ -877,3 +878,39 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+        except AttributeError:
+            return []
+
+    def get_rls_ids(self, table) -> List[int]:
```
add typing for args

altef replied:
Added

serenajiang reviewed (commented):
cooool

altef committed:
Catchup.

altef committed:
Merge branch 'row-level-security' of https://github.com/altef/incubator-superset into row-level-security

altef committed:
Changes as per @serenajiang.

altef committed:
Blacked.

altef committed:
Blacked and added some attributes to check for.

mistercrunch commented on superset/connectors/sqla/views.py:
```diff
@@ -221,6 +221,51 @@ class SqlMetricInlineView(CompactCRUDMixin, SupersetModelView):
     edit_form_extra_fields = add_form_extra_fields
 
 
+appbuilder.add_view_no_menu(SqlMetricInlineView)
```
Recently we merged a PR that moved all `add_view*` calls outside of view module scope and onto `superset/app.py`.

altef replied:
Oops, _removed_.

mistercrunch commented on superset/security/manager.py:
```diff
@@ -877,3 +878,34 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+        if (
+            hasattr(g, "user")
+            and hasattr(g.user, "roles")
+            and hasattr(table, "row_level_security")
```
Trying to make sense of all 3 `hasattr` here. So `g.user` should be defined if a user is logged in, and if defined, it seems like `user.roles` should always be defined, unless I'm missing something.

Now I'm guess `row_level_security` has to do with `DruidDatasource`. Should we just add a `row_level_security = []` as a class attribute there (or better, on `BaseDatasource`) for consistency?

altef replied:
Certain automated tests failed if i didn't include `hasattr(g.user, "roles")`.  I imagine it was `AttributeError: 'AnonymousUserMixin' object has no attribute 'id'`.

Since they're now retrieved by a query, a lot of this is no longer necessary.

mistercrunch reviewed (changes requested):
We need a good set of unit tests around this (it's a security/data-access feature!). I'm not sure how well the "Public" role use cases work at this time, and whether they are unit tested, but we may want to make sure this work for anonymous / Public users.

mistercrunch commented on superset/security/manager.py:
```diff
@@ -877,3 +878,34 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+            roles = [role.id for role in g.user.roles]
+            return [
+                f
+                for f in table.row_level_security_filters
```
Given that by default `sqlalchemy` lazy-loads relationships, this would emit 1+number_of_rls_in_table queries.

Let's write a single `sqlalchemy` query that returns the list of filters with the proper `IN` clause

altef replied:
Done - this reduces the number of necessary `hasattr` calls as well.

altef committed:
Changed to a manual query as per @mistercrunch.

altef committed:
Blacked.

altef committed:
Merge branch 'master' into row-level-security

altef committed:
Another migration in the meantime.

altef committed:
Black wanted some whitespace changes.

altef committed:
AttributeError: 'AnonymousUserMixin' object has no attribute 'id'.

altef committed:
Oops, did hasattr backwards.

axuew commented:
Different users see different data for the same role.In order to reduce the need for each user to configure a role separately, it is suggested to add a user field for filtering (the default value of the user field is ALL, so when selecting ALL, use role permissions directly; if configured to a single user, only take the permissions of the current user).

altef committed:
Merge branch 'master' into row-level-security

mistercrunch commented on superset/connectors/sqla/views.py:
```diff
@@ -225,6 +225,48 @@ class SqlMetricInlineView(CompactCRUDMixin, SupersetModelView):
+    }
+
+
+appbuilder.add_view(
```
Let's call this in `superset/app.py` in the `init_views` method

altef replied:
_moved_

mistercrunch commented on superset/security/manager.py:
```diff
@@ -887,3 +888,34 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+        if hasattr(g, "user") and hasattr(g.user, "id"):
+            from superset import db
+
+            result = db.session.execute(
```
It's the right query, but we should write this using SQLAlchemy SQL expressions

altef replied:
rewritten!

altef committed:
Merge branch 'master' into row-level-security

altef committed:
Changes as per @mistercrunch.

altef committed:
Doesn't look like text us required here anymore.

justin-barton commented:
@mistercrunch any chance that you could have another look at this? We've got a project for which we'd like to use this and we're just waiting on this merge so that we don't have to fork.

praneeth-m commented:
@mistercrunch will you be able to review this anytime soon? I'm considering using Superset on a project but can't use it without row-level security. Kindly let me know. Thank you.

dpgaspar commented on superset/connectors/sqla/views.py:
```diff
@@ -225,6 +225,38 @@ class SqlMetricInlineView(CompactCRUDMixin, SupersetModelView):
+    add_title = _("Add Row level security filter")
+    edit_title = _("Edit Row level security filter")
+
+    list_columns = ["table", "roles", "clause", "creator", "modified"]
```
small improvement: list_columns = ["table.table_name", "roles", "clause", "creator", "modified"]

dpgaspar commented on superset/connectors/sqla/views.py:
```diff
@@ -225,6 +225,38 @@ class SqlMetricInlineView(CompactCRUDMixin, SupersetModelView):
+    edit_title = _("Edit Row level security filter")
+
+    list_columns = ["table", "roles", "clause", "creator", "modified"]
+    order_columns = ["modified"]
```
small improvement: order_columns = ["table.table_name", "clause", "modified"]

dpgaspar reviewed (commented):
Gave it a run and looks good. Adding tests would be really important

villebro reviewed (changes requested):
Thanks for the review @dpgaspar ; while I initially felt it was ok to get this merged with "functionality only", given that this is a security feature, I don't feel comfortable merging this without at least some bare minimum tests. @altef if you need help with designing tests feel free to reach out on slack.

dpgaspar commented:
Note that: `RowLevelSecurityFilter` is only given by default to the `Admin` role and the `Admin` Role does have the all_datasources security permission. But, if users create a specific role with access to `RowLevelSecurityFilter` MVC and a custom datasource access, the `table` dropdown will not be correctly filtered by that custom datasource access. So we are assuming a default security config, a custom security config could potentially give access to setting filters on tables that users do not have access to.

altef commented:
@dpgaspar, @villebro; What test do you want?

villebro commented:
@altef having tests that validates that the query being generated correctly adds the expected where clause to the query based on a given role would be nice. We can add more tests later, but for now that should be enough.

altef committed:
Changes as per @dpgaspar

altef committed:
Two RLS tests.

altef commented:
@villebro - added

villebro reviewed (approved):
Thanks @altef ; This is looking pretty good to me 👍 Let's leave it open for a day or two for others to be able to chime in, and get this merged asap if no further objections come up.

villebro commented:
@mistercrunch @dpgaspar @etr2460 @serenajiang I think this is ready for merging; I'm hoping to merge this soon, so if there are any further objections please raise them in the next few days.

villebro commented:
@altef we discussed this with @dpgaspar , and came to the conclusion that it would be prudent to add this behind a feature flag in `config.py` (e.g. `ENABLE_ROW_LEVEL_SECURITY = False`) with the comment above raised by @dpgaspar (`Note that: RowLevelSecurityFilter...`), also adding a note in `UPDATING.md` on how to enable the feature.

I apologize for the long time to market on this feature, but we want to make sure no existing Superset deployment unknowingly opens up a potential security hole. Once this is done this is cleared for merging.

altef committed:
Row level security is now disabled by default via the feature flag ENABLE_ROW_LEVEL_SECURITY.

altef committed:
Merge branch 'master' into row-level-security

New db migration I guess

altef committed:
New head to revise.

altef commented:
@villebro - added!

villebro commented on superset/config.py:
```diff
@@ -723,6 +723,9 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
 }
 
+# Do you want to enable Row Level Security?
+ENABLE_ROW_LEVEL_SECURITY = False
```
Can we add the comment that @dpgaspar added earlier?

> Note that: RowLevelSecurityFilter is only given by default to the Admin role and the Admin Role does have the all_datasources security permission. But, if users create a specific role with access to RowLevelSecurityFilter MVC and a custom datasource access, the table dropdown will not be correctly filtered by that custom datasource access. So we are assuming a default security config, a custom security config could potentially give access to setting filters on tables that users do not have access to.

It doesn't need to be the exact same comment, but something that makes it possible for the admin enabling the feature to understand that there might be security implications that have yet to be fully addressed.

altef replied:
swapped

altef committed:
Changed the comment.

villebro commented:
Thanks @altef for quickly addressing the last change requests and your patience with the prolonged review process!

john-bodley commented on superset/security/manager.py:
```diff
@@ -891,3 +892,48 @@ def assert_viz_permission(self, viz: "BaseViz") -> None:
+        :returns: A list of IDs.
+        """
+        ids = [f.id for f in self.get_rls_filters(table)]
+        ids.sort()  # Combinations rather than permutations
```
It seems these IDs are sorted to satisfy the cache consistency. Rather than storing these as an ordered list why not return these as a set?

john-bodley commented on superset/connectors/sqla/models.py:
```diff
@@ -646,6 +646,19 @@ def adhoc_metric_to_sqla(self, metric: Dict, cols: Dict) -> Optional[Column]:
+        :rtype: List[str]
+        """
+        return [
+            text("({})".format(template_processor.process_template(f.clause)))
```
Could we use an `f` string here as opposed to `.format()`?

AaronCH5 commented:
Hi, I want to post our current processing ideas, but there are still two points that have not been realized. 
The interface is similar to @altef  and the main processing is filtering during drawing (including the user's regular Role and new permission at the line level). The operation interface is as follows:
1.![image](https://user-images.githubusercontent.com/24953071/76485414-7ec51900-6457-11ea-9e9f-64bf5fc0d55a.png)
2.![image](https://user-images.githubusercontent.com/24953071/76485427-88e71780-6457-11ea-99fa-f3f54e50eaf9.png)
3.![image](https://user-images.githubusercontent.com/24953071/76485542-c350b480-6457-11ea-8566-3e6153d79597.png)
The query interface after authorization is as follows：
4.![image](https://user-images.githubusercontent.com/24953071/76485627-06ab2300-6458-11ea-91b5-13f3d5922db2.png)

AaronCH5 commented:
@altef  Can U give some advice, the difference between the two treatment methods, as well as the feasibility, expectation.

AaronCH5 commented:
@altef I have a few questions:
1. Which version did you modify (we used 0.28.1)?
2. Your processing code is placed in viz.py-cache_key, how to ensure that the filtering is effective (I found that normal interaction processing is only in viz.py-query_obj)?
![image](https://user-images.githubusercontent.com/24953071/76512222-46413180-648f-11ea-8c1c-f59da12024e1.png)

altef commented:
Hi @AaronCH5,  I'm not completely certain what you're asking, but one difference is that in this pull request (which has been merged) instead of specifying a column and a value you specify a clause that will be added to the query.   So in your example above where you specify the column name `avg_circulate_duration` and the value `toB`, you would instead just write something like `avg_circulate_duration='toB'`.  

Since you can write whatever you want in there, this allows you to go pretty deep down the filter hole.  You can use multiple columns, or reference other tables as necessary.  For example, it would allow you to move complex permissions into its own table should you so desire:

> `avg_circulate_duration IN (SELECT allowed_avg_circulate_duration FROM custompermissions WHERE user_id='{{current_user_id()}}')`

The actual processing code where the filter clause gets added to the query is in https://github.com/apache/incubator-superset/blob/master/superset/connectors/sqla/models.py#L867

The cache key code in viz.py you referenced (https://github.com/apache/incubator-superset/blob/master/superset/viz.py#L394) adds the row level security IDs to the cache key, in order to differentiate cached query results for queries that would otherwise be the same but for different RLS rules.

AaronCH5 commented:
Hi @altef . First of all, thank you for your prompt answer. I can understand that your processing method has a higher degree of freedom and is more convenient; but I have not understood what you said 【in order to differentiate cached query results for queries that would otherwise be the same but for different RLS rules】. I tried your pull code in the local environment before, but found that filtering is not effective. In addition, I also want to know which version of the transformation you are based on, because I find your individual changes The file does not exist in the version I am using (0.28.1).
![image](https://user-images.githubusercontent.com/24953071/76593749-1f880700-6532-11ea-941c-7382b7895d1d.png)
![image](https://user-images.githubusercontent.com/24953071/76593760-244cbb00-6532-11ea-8aaf-a9872b1780ce.png)

durchgedreht commented:
Hi,
one question (that might be related): Is the filters applied on the fly also on logged in users or is a re-logon needed?

Grouping works for me, using 0.36.0rc1 
https://github.com/apache/incubator-superset/releases

Might be worth mentioning JINJA scripting works as well!  Thanks for this feature altef!

altef commented:
As far as I know they should be applied on-the-fly, and shouldn't require a re-login.  Maybe take that with a grain of salt though; I don't currently have a Superset environment running so I can't test it to verify.

altef commented:
@AaronCH5 It looks from a few posts past that you're using Druid.  The filtering occurs in `sqla/models.py` so it probably isn't applicable to Druid (I'm not familiar with Druid).  Looking briefly at the the `druid/models.py` file, there appears to be a function called [run_query](https://github.com/apache/incubator-superset/blob/master/superset/connectors/druid/models.py#L1137) - maybe you could add the filtering code there?  Again, I'm not familiar with Druid so I could be way off in that regard.

AaronCH5 commented:
@altef Thank you very much for your confusion, I will try again. Thanks

sahiljain001 commented:
Hi All,

We are looking for a solution in which we need to show a particular account_id data to a client. So when client1 logs in then he should be able to view his data only and when client2 logs in he should be able to view his data only. Is there any workaround in apache superset for the same as 0.36 branch is yet to be released.

Thanks

axelet commented:
@altef @villebro @justin-barton @mistercrunch Thank you for the update! It's a nice feature and we find it very use useful. 

Recently I was looking through the code and found a small issue with SELECT statement in the **get_rls_filters()** func. So I made this PR, could you pls take a look? https://github.com/apache/incubator-superset/pull/9365

altef commented:
@axelet hey, thanks for that! It looks good to me; I must have introduced that in one of the query format changes.

akushniarevich-gd commented:
Yeah, thanks for having a look :)

axelet commented:
Yeah, thanks for having a look :)

Aaron-Mhs commented:
> Hi All,
> 
> We are looking for a solution in which we need to show a particular account_id data to a client. So when client1 logs in then he should be able to view his data only and when client2 logs in he should be able to view his data only. Is there any workaround in apache superset for the same as 0.36 branch is yet to be released.
> 
> Thanks

Hi, you can try to look at this(https://github.com/apache/incubator-superset/issues/9320), it may be helpful to you, and it will take effect after configuration.

asen-aura commented:
Do these filters get applied dynamically to existing dashboards? For example, say I have transaction data for 100 tenants and I create a dashboard with a single big number chart that simply sums the transaction amount. Ideally, I want to expose (view access only) this single dashboard to all the tenants and rest assured that the big number chart each tenant sees in the dashboard is the sum of transactions made only by the viewing tenant.

axelet commented:
Hey, @asen-aura . I believe that filters are applied on the query level. So, yes, your tenants will be able to see only their data, as well as all the aggregations will be done only for the tenant data.

The feature basically adds WHERE clause to SQL query.
For, example:
SELECT * FROM your_table;
transforms to
SELECT * FROM your_table
WHERE tenant = 'foo';  

Regarding dashboards question all the filters applied for tables or slices (charts). So, if you change the filters for the dashboard's table or chart it will dynamically be applied to dashboard (as soon as you refresh the dashboard).

Mhs-Aaron commented:
@altef Hi, I have a question. At present, RLS is only applicable to Charts production, or it is also applicable to SQLLab. I am looking forward to your clarification.

altef commented:
hey @Mhs-Aaron,  I've never actually had reason to use SQLLab so I'm not clear on how it works.  I've mostly disabled permissions to it on my instance.  The RLS clause gets added in the [`get_sqla_query()` function](https://github.com/apache/incubator-superset/blob/e2f676445b406ddc3793949154270ab7a533a0e7/superset/connectors/sqla/models.py#L1103) and  I'm not sure if SQLLab makes use of that function or not (though I somewhat doubt it).

Mhs-Aaron commented:
@altef  Thank you very much for your timely feedback, I would like to know what kind of user is appropriate for the SQLLab menu assignment, if it is open to business people, then the same RLS problem will be involved.I wonder if you have given any thought to this point.
In addition, there is a new function which I am interested in [.](https://github.com/apache/incubator-superset/issues/12203)

axelet commented:
@Mhs-Aaron Hi, regarding RLS filters I don't think they have effect on Sqllab. SQLLab is more like an admin tool imo. But things could change since I visited the code last time

shenrie commented:
@altef A very useful feature, so thank you for implementing.  I do have a question that I am confused about. The documentation and parts of the code reference "table and roles", however, in the UI it seems that it is not actually a database table that needs to be configured in the RLS, but actually a dataset. Is that a true or am I misunderstanding?

The reason that I ask is in a scenario where I have a dozen different dashboards that use their own unique datasets, but which all include the same database table "abc" as part of their query, I would like the same RLS rule to apply to all of these dashboards, it would be great if I could specify the table "abc" in the rule, since that single table is common across all datasets. However, it would seem from my review of the UI/code that it is not the actual database table name that is defined in the RLS rule, but that all dozen datasets would have to be specified, which is not very scalable. 

Am I understanding this feature correctly? Or is there some other way to do what I would like to do by specify a single common table that is used within multiple dashboard datasets to ensure the RLS is enforced across all of them?  

Thank you in advance.

altef commented:
@shenrie If I recall correctly (and this may be out-of-date), it's on a table _but not by table name_ - a [table](https://github.com/altef/incubator-superset/blob/5eeb296223dbd25e0fbca8386faee152e4402499/superset/connectors/sqla/models.py#L341) entry in superset, which has a reference to a specific [database](https://github.com/apache/superset/blob/master/superset/models/core.py#L109).

I've had luck doing that sort of thing from outside superset. For example, if I had a script that added the database and tables to superset, I might have it set the RLS rules for any table called `users`.  This of course doesn't help at all if you're adding them manually.  Thankfully I was able to settle on a simple, consistent rule that could be set and never adjusted by restricting data by user ID based on a convention I maintain on the DB side.  It was in the vein of :

```sql
data_group_id IN (
	SELECT data_group_id FROM users_to_data_groups 
	WHERE user_id={{ cache_key_wrapper(current_user_id()) }}
)
```
